### PR TITLE
simplify DAGGER modifiers in operators

### PR DESCRIPTION
**Pull request by stylewarning**
This will simplify modifier chains such as:

    DAGGER CONTROLLED DAGGER FORKED F(...) ...

by bringing DAGGERs to the front and canceling them out.

This commit also simplifies parameter arithmetic by eliding factors of
one.

This also allows programs like

```
DAGGER DAGGER CPHASE(x) 0 1
```

to compile.

In order to accomplish this, we re-work where some of the simplification and the like happens.

### CODE REVIEW

**Comment by stylewarning** [comment]
Tests are "passing", but the golden test files are failing. I'll have to think about how I want to deal with this. Options are:

- Re-generate the golden tests.
- Don't apply the simplification transformation post-parsing, and let the golden tests *just* test raw parsing and nothing else.

**Review by ecpeterson — APPROVED** [review]
I don’t object to any of this, but I wonder if I’d prefer applying individual compilers to the innermost (few) modifiers of a named operator to canonicalization pass, like WHATEVERn … WHATEVER1 DAGGER CPHASE(t) q -> WHATEVERn … WHATEVER1 CPHASE(-t) q.

Fiasco is unhappy tho.

**Comment by stylewarning** [comment]
from Quil meeting, there is general agreement to move optimization/simplification transforms to compiler-hook, possibly needing to regenerate the golden test files

**Comment by stylewarning** [comment]
@kilimanjaro @ecpeterson @notmgsk this is ready for review now

File: src/analysis/simplify-arithmetic.lisp [hunk 1]
```diff
@@ -159,10 +157,17 @@ AFFINE-REPRESENTATION                       (* 2.0 theta[0])
 "
   (let ((expr nil))
     (dohash ((ref coefficient) (affine-representation-coefficients rep))
-      (unless (double= 0 coefficient)
-        (setf expr (if expr
+      (cond
+        ((double= 0 coefficient)        ; skip
+         nil)
+        ((double= 1 coefficient)        ; don't add '1 *` factor
```
Comment (notmgsk) [inline src/analysis/simplify-arithmetic.lisp:163]: chef's kiss

**Review by notmgsk — APPROVED** [review]

**Pull request merged**